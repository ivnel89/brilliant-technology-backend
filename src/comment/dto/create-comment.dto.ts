import { ApiProperty } from "@nestjs/swagger";
import { IsOptional, IsString, IsUUID, Length } from "class-validator";

export class CreateCommentDto {
  @ApiProperty()
  @IsString()
  @Length(1, 10000)
  content: string;

  @ApiProperty()
  @IsString()
  @IsUUID()
  authorId: string;

  @ApiProperty()
  @IsString()
  @IsUUID()
  articleId: string;

  @ApiProperty({
    nullable: true,
    required: false
  })
  @IsString()
  @IsUUID()
  @IsOptional()
  parentCommentId: string;
}
