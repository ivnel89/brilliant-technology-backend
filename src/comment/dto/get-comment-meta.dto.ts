import { ApiProperty } from "@nestjs/swagger";
import { IsArray, IsString, IsUUID } from "class-validator";

export class GetCommentsDto {
  @ApiProperty()
  @IsString()
  @IsUUID()
  userId: string;

  @ApiProperty({
     isArray: true
  })
  @IsArray()
  commentIds: Array<string>;
}
