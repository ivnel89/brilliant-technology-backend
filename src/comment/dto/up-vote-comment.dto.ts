import { ApiProperty } from "@nestjs/swagger";
import { IsString, IsUUID } from "class-validator";

export class UpVoteCommentDto {
  @ApiProperty()
  @IsString()
  @IsUUID()
  userId: string;
}
