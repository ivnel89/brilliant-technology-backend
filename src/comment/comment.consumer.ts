import { InjectQueue, OnQueueActive, Process, Processor } from '@nestjs/bull';
import { Injectable } from '@nestjs/common';
import { Job, Queue } from 'bull';
import { QueueName } from 'src/config/queueName.enum';
import { CommentService } from './comment.service';
import { UpVoteCommentDto } from './dto/up-vote-comment.dto';

@Injectable()
@Processor(QueueName.COMMENTS_QUEUE)
export class CommentConsumer {
  constructor(
    private readonly commentService: CommentService,
  ) {}

  @OnQueueActive()
  onActive(job: Job) {
    console.log(
      `Processing job ${job.id} of type ${job.name} with data ${job.data}...`,
    );
  }

  @Process('addUpVote')
  async processAddUpVote(job: Job<{
    commentId: string,
    upVoteCommentDto: UpVoteCommentDto
  }>){
    const {
      data: { commentId, upVoteCommentDto },
    } = job;
    console.log('consuming job')
    this.commentService.addUpVote(commentId, upVoteCommentDto)
  }

  @Process('removeUpVote')
  async processRemoveUpVote(job: Job<{
    commentId: string,
    upVoteCommentDto: UpVoteCommentDto
  }>){
    const {
        data: { commentId, upVoteCommentDto },
      } = job;
      console.log('consuming job')
      this.commentService.removeUpVote(commentId, upVoteCommentDto)
  }
  
}
