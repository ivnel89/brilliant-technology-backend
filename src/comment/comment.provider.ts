import { InjectQueue } from '@nestjs/bull';
import { Injectable } from '@nestjs/common';
import { Queue } from 'bull';
import { QueueName } from 'src/config/queueName.enum';
import { UpVoteCommentDto } from './dto/up-vote-comment.dto';

@Injectable()
export class CommentProvider {
  constructor(
    @InjectQueue(QueueName.COMMENTS_QUEUE) private commentsQueue: Queue
  ) {}

  async addUpVoteJob(commentId: string, upVoteCommentDto: UpVoteCommentDto){
    console.log('adding job')
    await this.commentsQueue.add('addUpVote',{
      commentId,
      upVoteCommentDto
    })
  }

  async removeUpVoteJob(commentId: string, upVoteCommentDto: UpVoteCommentDto){
    console.log('adding job')
    await this.commentsQueue.add('removeUpVote', {
      commentId,
      upVoteCommentDto,
    });
  }
  
}
