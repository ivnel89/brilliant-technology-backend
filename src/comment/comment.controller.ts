import { Controller, Post, Body, Put, Param, Get, Query, } from '@nestjs/common';
import { CommentConsumer } from './comment.consumer';
import { CommentProvider } from './comment.provider';
import { CommentService } from './comment.service';
import { CommentContract } from './contract/comment.contract';
import { CreateCommentDto } from './dto/create-comment.dto';
import { GetCommentsDto } from './dto/get-comment-meta.dto';
import { UpVoteCommentDto } from './dto/up-vote-comment.dto';
import { Comment } from './entities/comment.entity';
import { CommentContractMapper } from './mapper/comment.mapper';

@Controller('comment')
export class CommentController {
  constructor(
    private readonly commentService: CommentService,
    private readonly commentProvider: CommentProvider
    ) {}
  private commentContractMapper = new CommentContractMapper();

  @Post()
  async create(
    @Body() createCommentDto: CreateCommentDto,
  ): Promise<CommentContract> {
    const comment = await this.commentService.create(createCommentDto);
    return this.commentContractMapper.build(comment);
  }

  @Put(':id/upvote/add')
  async addUpVote(
    @Param('id') id: string,
    @Body() upVoteCommentDto: UpVoteCommentDto,
  ): Promise<CommentContract> {
    await this.commentProvider.addUpVoteJob(id, upVoteCommentDto);
    const comment = await this.commentService.findOne(id);
    return this.commentContractMapper.build(comment, upVoteCommentDto.userId);
  }

  @Put(':id/upvote/remove')
  async removeUpVote(
    @Param('id') id: string,
    @Body() upVoteCommentDto: UpVoteCommentDto,
  ): Promise<CommentContract> {
   await this.commentProvider.removeUpVoteJob(id, upVoteCommentDto);
    const comment = await this.commentService.findOne(id);
    return this.commentContractMapper.build(comment, upVoteCommentDto.userId);
  }

  @Get('')
  async getComments(
    @Query() getCommentMetaDto: GetCommentsDto,
  ): Promise<Array<CommentContract>>{
    let comments: Array<Comment>;
    if(getCommentMetaDto.articleId){
      comments = await this.commentService.getCommentsByArticleId(
        getCommentMetaDto.articleId,
      );
    }else{
      comments = await this.commentService.getComments(
        getCommentMetaDto.commentIds,
      );
    }
    return this.commentContractMapper.buildArray(
      comments,
      getCommentMetaDto.userId,
    );
  }
}
