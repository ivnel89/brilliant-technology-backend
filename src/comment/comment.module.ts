import { Module } from '@nestjs/common';
import { CommentService } from './comment.service';
import { CommentController } from './comment.controller';
import { UserModule } from 'src/user/user.module';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Comment } from './entities/comment.entity';
import { ArticleModule } from 'src/article/article.module';
import { CommentProvider } from './comment.provider';
import { BullModule } from '@nestjs/bull';
import { QueueName } from 'src/config/queueName.enum';
import { CommentConsumer } from './comment.consumer';

@Module({
  imports: [
    TypeOrmModule.forFeature([Comment]),
    UserModule,
    ArticleModule,
    BullModule.registerQueue(
      {
        name: QueueName.COMMENTS_QUEUE,
      },
    ),
  ],
  controllers: [CommentController],
  providers: [CommentService, CommentProvider, CommentConsumer],
})
export class CommentModule {}
