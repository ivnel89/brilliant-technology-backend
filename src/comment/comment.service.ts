import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { ArticleService } from 'src/article/article.service';
import { UserService } from 'src/user/user.service';
import { Repository } from 'typeorm';
import { CreateCommentDto } from './dto/create-comment.dto';
import { Comment } from './entities/comment.entity';

@Injectable()
export class CommentService {
  constructor(
    @InjectRepository(Comment)
    private commentRepository: Repository<Comment>,
    private readonly userService: UserService,
    private readonly articleService: ArticleService,
  ) {}

  async create(createCommentDto: CreateCommentDto) {
    const [ 
      user, 
      article
     ] = await Promise.all([
       this.userService.findOne(createCommentDto?.authorId),
       this.articleService.findOne(createCommentDto?.articleId)
      ])
    return this.commentRepository.save(
      new Comment(user, article, createCommentDto?.content),
    );
  }

}
