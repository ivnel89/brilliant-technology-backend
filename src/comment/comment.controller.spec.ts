import { Test, TestingModule } from '@nestjs/testing';
import { ArticleService } from 'src/article/article.service';
import { ArticleRepository } from 'src/article/test/article.repository.mock';
import { UserRepository } from 'src/user/test/user.repository.mock';
import { UserService } from 'src/user/user.service';
import { CommentController } from './comment.controller';
import { CommentService } from './comment.service';
import { CommentRepository } from './test/comment.repository.mock';
import { createCommentDto, mockCommentId, upVoteCommentDto } from './test/comment.dto.mock';
import { CreateCommentDto } from './dto/create-comment.dto';
import { validate } from 'class-validator';
import { plainToInstance } from 'class-transformer';
import { UpVoteCommentDto } from './dto/up-vote-comment.dto';
import { CommentConsumer } from './comment.consumer';
import { CommentProvider } from './comment.provider';
import { BullModule } from '@nestjs/bull';
import { QueueName } from 'src/config/queueName.enum';

describe('CommentController', () => {
  let controller: CommentController;
  let service: CommentService;
  let provider: CommentProvider;
  let serviceCreateSpy,
    serviceAddUpVoteSpy,
    serviceRemoveUpVoteSpy,
    providerAddUpVoteJobSpy,
    providerRemoveUpVoteJobSpy;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports:[
        BullModule.registerQueue({
          name: QueueName.COMMENTS_QUEUE
    })
      ],
      controllers: [CommentController],
      providers: [
        CommentService,
        CommentRepository,
        UserService,
        UserRepository,
        ArticleService,
        ArticleRepository,
        CommentConsumer,
        CommentProvider
      ],
    }).compile();

    controller = module.get<CommentController>(CommentController);
    service = module.get<CommentService>(CommentService);
    provider = module.get<CommentProvider>(CommentProvider);
    serviceCreateSpy = jest.spyOn(service, 'create');
    serviceAddUpVoteSpy = jest.spyOn(service, 'addUpVote');
    serviceRemoveUpVoteSpy = jest.spyOn(service, 'removeUpVote');
    providerAddUpVoteJobSpy = jest.spyOn(provider, 'addUpVoteJob');
    providerRemoveUpVoteJobSpy = jest.spyOn(provider, 'removeUpVoteJob');
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  describe('create', () => {
    const serviceCreateError = "serviceCreateError";
    it('should call create from commentService with right params', async () => {
      await expect(controller.create(createCommentDto)).resolves.toMatchSnapshot().then(() => {
        expect(serviceCreateSpy).toBeCalledWith(createCommentDto);
      })
    })
    it('should validate createCommentDto', async () => { 
      const invalidInput = new CreateCommentDto();
      invalidInput.articleId = 999 as any;
      invalidInput.authorId = "notUUID";
      invalidInput.content = undefined;
      const errors = await validate(plainToInstance(CreateCommentDto,invalidInput))
      expect(errors.length).toBe(3)
    })
    it('should throw an error when create from commentService throws an error', async () => {
      serviceCreateSpy.mockRejectedValueOnce(serviceCreateError);
      await expect(controller.create(createCommentDto)).rejects.toBe(
        serviceCreateError,
      );
    })
  })

  describe('upVote', () => {
    const serviceUpVoteError = "serviceUpVoteError";
    it('should call addUpVoteJob from commentProvider with right params', async () => {
      await expect(controller.addUpVote(mockCommentId,upVoteCommentDto)).resolves.toMatchSnapshot().then(() => {
        expect(providerAddUpVoteJobSpy).toBeCalledWith(mockCommentId,upVoteCommentDto);
        expect(serviceAddUpVoteSpy).not.toBeCalled();
      })
    })
    it('should validate upVoteCommentDto', async () => { 
      const invalidInput = new UpVoteCommentDto();
      invalidInput.userId = 999 as any;
      const errors = await validate(plainToInstance(UpVoteCommentDto,invalidInput))
      expect(errors.length).toBe(1)
    })
    it('should throw an error when upvoteJob from commentProvider throws an error', async () => {
      providerAddUpVoteJobSpy.mockRejectedValueOnce(serviceUpVoteError);
      await expect(
        controller.addUpVote(mockCommentId, upVoteCommentDto),
      ).rejects.toBe(serviceUpVoteError);
    })
  })

  describe('removeVote', () => {
    const serviceRemoveVoteError = "serviceRemoveVoteError";
    it('should call removeUpvoteJob from commentProvider with right params', async () => {
      await expect(controller.removeUpVote(mockCommentId,upVoteCommentDto)).resolves.toMatchSnapshot().then(() => {
        expect(providerRemoveUpVoteJobSpy).toBeCalledWith(mockCommentId,upVoteCommentDto);
        expect(serviceRemoveUpVoteSpy).not.toBeCalled();
      })
    })
    it('should validate upVoteCommentDto', async () => { 
      const invalidInput = new UpVoteCommentDto();
      invalidInput.userId = 999 as any;
      const errors = await validate(plainToInstance(UpVoteCommentDto,invalidInput))
      expect(errors.length).toBe(1)
    })
    it('should throw an error when removeUpvoteJob from commentProvider throws an error', async () => {
      providerRemoveUpVoteJobSpy.mockRejectedValueOnce(serviceRemoveVoteError);
      await expect(
        controller.removeUpVote(mockCommentId, upVoteCommentDto),
      ).rejects.toBe(serviceRemoveVoteError);
    })
  })

});
