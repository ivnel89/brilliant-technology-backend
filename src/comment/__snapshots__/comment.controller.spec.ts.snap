// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentController create should call create from commentService with right params 1`] = `
CommentContract {
  "author": UserContract {
    "displayPicture": "www.example.com/img.png",
    "firstName": "Foo",
    "id": undefined,
    "lastName": "Bar",
  },
  "content": "comment content",
  "createdDate": undefined,
  "id": undefined,
}
`;

exports[`CommentController removeVote should call removeUpvoteJob from commentProvider with right params 1`] = `
CommentContract {
  "author": UserContract {
    "displayPicture": "www.example.com/img.png",
    "firstName": "Foo",
    "id": undefined,
    "lastName": "Bar",
  },
  "content": "comment content",
  "createdDate": undefined,
  "id": undefined,
  "upVoted": false,
  "upVotes": 0,
}
`;

exports[`CommentController upVote should call addUpVoteJob from commentProvider with right params 1`] = `
CommentContract {
  "author": UserContract {
    "displayPicture": "www.example.com/img.png",
    "firstName": "Foo",
    "id": undefined,
    "lastName": "Bar",
  },
  "content": "comment content",
  "createdDate": undefined,
  "id": undefined,
  "upVoted": false,
  "upVotes": 0,
}
`;
